File: carcassone_training_platform/model/tile.py
from typing import List, Literal, Union
from dataclasses import dataclass

EdgeList = List[int]

@dataclass
class Feature:
    type: Literal["FieldFeature", "RoadFeature", "CityFeature"]
    connectedEdges: EdgeList

@dataclass
class FieldFeature(Feature):
    def __init__(self, connectedEdges: EdgeList):
        super().__init__("FieldFeature", connectedEdges)

@dataclass
class RoadFeature(Feature):
    hasInn: bool

    def __init__(self, connectedEdges: EdgeList, hasInn: bool):
        super().__init__("RoadFeature", connectedEdges)
        self.hasInn = hasInn

@dataclass
class CityFeature(Feature):
    hasShield: bool
    hasDouble: bool

    def __init__(self, connectedEdges: EdgeList, hasShield: bool, hasDouble: bool):
        super().__init__("CityFeature", connectedEdges)
        self.hasShield = hasShield
        self.hasDouble = hasDouble

@dataclass
class GardenFeature(Feature):

    def __init__(self, connectedEdges: EdgeList = []):
        super().__init__("GardenFeature", connectedEdges)

@dataclass
class Tile:
    id: str
    image: str
    features: List[Feature]

File: carcassone_frontend/src/types/Tile.ts
// 前端编辑用
export interface FeatureEditor {
    id: number;
    type: '草坪' | '城' | '路' | '修道院';
    connectedEdges: number[];
    hasShield?: boolean;
    hasDouble?: boolean;
    hasInn?: boolean;
}

export interface TileEditor {
    id: number;
    features: FeatureEditor[];
    image?: string;
}

// 提交给后端的格式
export interface FeaturePayload {
    id: string;
    name: string;
    description: string;
}

export interface TilePayload {
    id: number;
    image: string;
    features: FeaturePayload[];
}
